src/libraries/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -751,7 +751,7 @@ public abstract partial class JsonConverter<T> : System.Text.Json.Serialization.
         protected internal JsonConverter() { }
         public override bool CanConvert(System.Type typeToConvert) { throw null; }
         public abstract T Read(ref System.Text.Json.Utf8JsonReader reader, System.Type typeToConvert, System.Text.Json.JsonSerializerOptions options);
-        public abstract void Write(System.Text.Json.Utf8JsonWriter writer, T value, System.Text.Json.JsonSerializerOptions options);
+        public abstract void Write(System.Text.Json.Utf8JsonWriter writer, [System.Diagnostics.CodeAnalysis.NotNull] T value, System.Text.Json.JsonSerializerOptions options);
```

Comment:
NotNull doesn't make sense here. Did you mean DisallowNull? And just so I understand, it's invalid to write nulls but read can return nulls?

Reply:
> And just so I understand, it's invalid to write nulls but read can return nulls?

Yes. `Write` won't receive null but `Read` can return null.

The intent is the `JsonSerializer` will never pass in null T values to the `JsonConverter<T>.Write` method and hence the implementer of that method doesn't need to do a null check on it.

They can certainly continue to write null JSON literals in the method itself or do whatever else. If an nullable object has the `null` value, the serializer writes null for you (and doesn't call the `JsonConverter`).

On the read side, the implementer should honor the nullability of the T. If T is nullable, the implementer of `JsonConverter<T>.Read` can certainly return null if they wish. If T is non-nullable (like string, or value type), it doesn't make sense for them to do so.

For types where null doesn't make sense (i.e. non-nullable valuetypes), the `JsonSerializer` can and will pass in a `Utf8JsonReader` with the `null` TokenType if the payload we are processing has the null JSON literal.
For types where null is allowed (nullable/non-nullable reference types), the `JsonSerializer` will **not** pass in a `Utf8JsonReader` in that state and will eagerly set the object to null up front, without calling the `JsonConverter`.

```C#
#nullable enable

using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using Xunit;

namespace System.Text.Json.Serialization.Tests
{
    public static partial class PropertyNameTests
    {
        [Fact]
        public static void QuickTest()
        {
            var opts = new JsonSerializerOptions
            {
                Converters =
                {
                    new MyStringConverter(),
                    new MyNullableStringConverter(),
                    new MyNullableIntConverter(),
                    new MyIntConverter()
                }
            };

            string json = "{\"foo\": null, \"bar\": null, \"baz\": null,\"nullBaz\": null}";

            MyClass output = JsonSerializer.Deserialize<MyClass>(json, opts)!;
            Assert.Null(output.foo);
            Assert.Null(output.bar);
            Assert.Equal(0, output.baz);
            Assert.Null(output.nullBaz);
        }

        public class MyClass
        {
#pragma warning disable CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
            public string foo { get; set; }
#pragma warning restore CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.

            public string? bar { get; set; }

            public int baz { get; set; }

            public int? nullBaz { get; set; }
        }

        public class MyStringConverter : JsonConverter<string>
        {
            public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                Debug.Assert(reader.TokenType != JsonTokenType.Null);

                return reader.GetString()!;
            }

            public override void Write(Utf8JsonWriter writer, [DisallowNull] string value, JsonSerializerOptions options)
            {
                throw new NotImplementedException();
            }
        }

        public class MyNullableStringConverter : JsonConverter<string?>
        {
            public override string? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                Debug.Assert(reader.TokenType != JsonTokenType.Null);

                return reader.GetString();
            }

            public override void Write(Utf8JsonWriter writer, [DisallowNull] string? value, JsonSerializerOptions options)
            {
                throw new NotImplementedException();
            }
        }

        public class MyIntConverter : JsonConverter<int>
        {
            public override int Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                if (reader.TokenType == JsonTokenType.Null)
                    return default;
                return reader.GetInt32();
            }

            public override void Write(Utf8JsonWriter writer, [DisallowNull] int value, JsonSerializerOptions options)
            {
                throw new NotImplementedException();
            }
        }

        public class MyNullableIntConverter : JsonConverter<int?>
        {
            public override int? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
            {
                if (reader.TokenType == JsonTokenType.Null)
                    return null;
                return reader.GetInt32();
            }

            public override void Write(Utf8JsonWriter writer, [DisallowNull] int? value, JsonSerializerOptions options)
            {
                throw new NotImplementedException();
            }
        }
    }
}
```